lib/src/signing.rs

```diff
@@ -0,0 +1,230 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::RwLock;
+
+use thiserror::Error;
+
+use crate::backend::CommitId;
+use crate::settings::UserSettings;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SigStatus {
+    /// Valid signature that matches the data.
+    Good,
+    /// Valid signature that could not be verified (e.g. due to an unknown key).
+    Unknown,
+    /// Valid signature that does not match the signed data.
+    Bad,
+    /// Invalid signature.
+    Invalid,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Verification {
+    /// The status of the signature.
+    pub status: SigStatus,
+    /// The key representation, if available. For GPG, this will be the key
+    /// fingerprint.
+    pub key: Option<String>,
+    /// A display string, if available. For GPG, this will be formatted primary
+    /// user ID.
+    pub display: Option<String>,
+}
+
+impl Verification {
+    pub fn unknown() -> Self {
+        Self {
+            status: SigStatus::Unknown,
+            key: None,
+            display: None,
+        }
+    }
+    pub fn invalid() -> Self {
+        Self {
+            status: SigStatus::Invalid,
+            key: None,
+            display: None,
+        }
+    }
+}
+
+/// The backend for signing and verifying cryptographic signatures.
+///
+/// This allows using different signers, such as GPG or SSH, or different
+/// versions of them.
+pub trait SigningBackend: Debug + Send + Sync {
+    /// Check if the signature was created by this backend implementation.
+    ///
+    /// Should check the signature format, usually just looks at the prefix.
+    fn is_of_this_type(&self, signature: &[u8]) -> bool;
+
+    /// Create a signature for arbitrary data.
+    ///
+    /// The `key` parameter is what `jj sign` receives as key argument, or what
+    /// is configured in the `signing.key` config.
+    fn sign(&self, data: &[u8], key: Option<&str>) -> SignResult<Vec<u8>>;
+
+    /// Verify a signature. Should be reflexive with `sign`:
+    /// ```rust,ignore
+    /// verify(data, sign(data)?)?.status == SigStatus::Good
+    /// ```
+    fn verify(&self, data: &[u8], signature: &[u8]) -> SignResult<Verification>;
+}
+
+#[derive(Debug, Error)]
+#[error("Signing error: {0}")]
+pub struct SignError(#[from] Box<dyn std::error::Error + Send + Sync>);
+
+pub type SignResult<T> = Result<T, SignError>;
+
+#[derive(Debug, Error)]
+pub enum SignInitError {
+    #[error("Unknown signing backend configured: {0}")]
+    UnknownBackend(String),
+    #[error("Failed to initialize signing: {0}")]
+    BackendError(Box<dyn std::error::Error + Send + Sync>),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum SignBehavior {
```

Comment:
nit: Should we call this `RewriteBehavior` to clarify that it's behavior related to rewriting commits (right?)?

Reply:
Well, it's both
I think not having the word `sign` makes the name more confusing, esp if you encounter it like in the commit builder, `set_rewrite_behavior` could mean whatever, rewriting commits is like the main function of jj :)
OTOH I agree that this type bears more meaning in relation to rewrites
OTOOH Signing is kind of stateless, this type helps decide "do we sign this commit X or do we not", it does not think about rewrites per se